backend/db/sql-generic.h

```diff
@@ -0,0 +1,1174 @@
+	return TRUE;
+_error:
+	g_string_free(sql, TRUE);
+	if (schema)
+	{
+		if (schema_initialized)
+			bson_destroy(schema);
+		g_free(schema);
+	}
+	if (variables_index)
+		g_hash_table_destroy(variables_index);
+	if (variables_type)
+		g_hash_table_destroy(variables_type);
+	return FALSE;
+}
+static gboolean
+backend_iterate(gpointer _iterator, bson_t* metadata, GError** error)
+{
+	const char* name;
+	guint i;
+	JDBType type;
+	gint ret;
+	JSqlCacheSQLPrepared* prepared = _iterator;
+	gboolean found = FALSE;
+	j_sql_step(prepared->stmt, ret)
+	{
+		found = TRUE;
+		for (i = 0; i < prepared->variables_count; i++)
+		{
+			name = g_hash_table_lookup(prepared->variables_index, GINT_TO_POINTER(i));
+			type = GPOINTER_TO_INT(g_hash_table_lookup(prepared->variables_type, name));
+			switch (type)
+			{
+			case J_DB_TYPE_SINT32:
+				ret = bson_append_int32(metadata, name, -1, j_sql_column_sint32(prepared->stmt, i));
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "SINT32");
+				break;
+			case J_DB_TYPE_UINT32:
+				ret = bson_append_int32(metadata, name, -1, j_sql_column_uint32(prepared->stmt, i));
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "UINT32");
+				break;
+			case J_DB_TYPE_FLOAT32:
+				ret = bson_append_double(metadata, name, -1, j_sql_column_float32(prepared->stmt, i));
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "FLOAT32");
+				break;
+			case J_DB_TYPE_SINT64:
+				ret = bson_append_int64(metadata, name, -1, j_sql_column_sint64(prepared->stmt, i));
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "SINT64");
+				break;
+			case J_DB_TYPE_UINT64:
+				ret = bson_append_int64(metadata, name, -1, j_sql_column_uint64(prepared->stmt, i));
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "UINT64");
+				break;
+			case J_DB_TYPE_FLOAT64:
+				ret = bson_append_double(metadata, name, -1, j_sql_column_float64(prepared->stmt, i));
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "FLOAT64");
+				break;
+			case J_DB_TYPE_STRING:
+				ret = bson_append_utf8(metadata, name, -1, j_sql_column_text(prepared->stmt, i), -1);
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "STRING");
+				break;
+			case J_DB_TYPE_BLOB:
+				if (j_sql_column_blob(prepared->stmt, i) != NULL)
+					ret = bson_append_binary(metadata, name, -1, BSON_SUBTYPE_BINARY, (const uint8_t*)j_sql_column_blob(prepared->stmt, i), j_sql_column_blob_len(prepared->stmt, i));
+				else
+					ret = bson_append_null(metadata, name, -1);
+				j_goto_error_backend(!ret, JULEA_BACKEND_ERROR_BSON_APPEND_FAILED, "BLOB");
+				break;
+			case _J_DB_TYPE_COUNT:
+			default:
+				j_goto_error_backend(TRUE, JULEA_BACKEND_ERROR_DB_TYPE_INVALID, "");
+			}
+		}
+	}
+	j_goto_error_backend(!found, JULEA_BACKEND_ERROR_ITERATOR_NO_MORE_ELEMENTS, "");
+	return TRUE;
+_error:
+	j_sql_reset(prepared->stmt);
+	return FALSE;
+}
+#endif
```

Comment:
Why is this code in a header? Is there a reason not to put it into a source file?

Reply:
The idea is, that new backends like for example mysql could define the sql* Macros such that instead of the sqlite-lib the mysql-lib is called. If this file would be a source-file it would be required that there are lots of function pointers in the function parameters - instead of the sql* Macros.